crates/core/tedge_agent/src/entity_manager/server.rs

```diff
@@ -0,0 +1,157 @@
+use async_trait::async_trait;
+use tedge_actors::LoggingSender;
+use tedge_actors::MessageSink;
+use tedge_actors::Sender;
+use tedge_actors::Server;
+use tedge_api::entity_store;
+use tedge_api::entity_store::EntityMetadata;
+use tedge_api::entity_store::EntityRegistrationMessage;
+use tedge_api::mqtt_topics::Channel;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::pending_entity_store::PendingEntityData;
+use tedge_api::EntityStore;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::TopicFilter;
+use tracing::error;
+
+#[derive(Debug)]
+pub enum EntityStoreRequest {
+    Get(EntityTopicId),
+    Create(EntityRegistrationMessage),
+    Delete(EntityTopicId),
+    MqttMessage(MqttMessage),
+}
+
+#[derive(Debug)]
+pub enum EntityStoreResponse {
+    Get(Option<EntityMetadata>),
+    Create(Result<(Vec<EntityTopicId>, Vec<PendingEntityData>), entity_store::Error>),
+    Delete(Vec<EntityTopicId>),
+    Ok,
+}
+
+pub struct EntityStoreServer {
+    entity_store: EntityStore,
+    mqtt_schema: MqttSchema,
+    mqtt_publisher: LoggingSender<MqttMessage>,
+    entity_auto_register: bool,
+}
+
+impl EntityStoreServer {
+    pub fn new(
+        entity_store: EntityStore,
+        mqtt_schema: MqttSchema,
+        mqtt_actor: &mut impl MessageSink<MqttMessage>,
+        entity_auto_register: bool,
+    ) -> Self {
+        let mqtt_publisher = LoggingSender::new("MqttPublisher".into(), mqtt_actor.get_sender());
+
+        Self {
+            entity_store,
+            mqtt_schema,
+            mqtt_publisher,
+            entity_auto_register,
+        }
+    }
+}
+
+#[async_trait]
+impl Server for EntityStoreServer {
+    type Request = EntityStoreRequest;
+    type Response = EntityStoreResponse;
+
+    fn name(&self) -> &str {
+        "ConcurrentWorker"
+    }
+
+    async fn handle(&mut self, request: EntityStoreRequest) -> EntityStoreResponse {
+        match request {
+            EntityStoreRequest::Get(topic_id) => {
+                let entity = self.entity_store.get(&topic_id);
+                EntityStoreResponse::Get(entity.cloned())
+            }
+            EntityStoreRequest::Create(entity) => {
+                let res = self.entity_store.update(entity);
+                EntityStoreResponse::Create(res)
+            }
+            EntityStoreRequest::Delete(topic_id) => {
+                let deleted_entities = self.entity_store.deregister_entity(&topic_id);
+                EntityStoreResponse::Delete(deleted_entities)
+            }
+            EntityStoreRequest::MqttMessage(mqtt_message) => {
+                self.process_mqtt_message(mqtt_message).await;
```

Comment:
I'm surprised that there is a lot more to do when a message is received from MQTT vs HTTP.

Why that?

Reply:
The difference is in the "pending entities" caching aspect. When the registration is done over MQTT, due to the lack of feedback to the client, we are forced to cache any early child reg messages, until the parents are also registered. But, with the HTTP, when a client tries to register before its parent, we can always immediately send the response back that the parent is not available so that the client can retry later. We don't **need to** do caching in this case. But, we **can** if we want to.